language_service/src/compilation.rs

```diff
@@ -113,21 +113,81 @@ impl Compilation {
+
+        let (sources, dependencies, store) = match &project {
+            Some(p) if p.errors.is_empty() => {
+                trace!("using buildable program from project");
+                let buildable_program =
+                    prepare_package_store(target_profile.into(), p.package_graph_sources.clone());
+
+                if !buildable_program.dependency_errors.is_empty() {
+                    return Self {
+                        package_store: buildable_program.store,
+                        user_package_id: 2.into(), // We don't have a legitimate user package id here, so fall back to a guess.
+                        compile_errors: buildable_program.dependency_errors,
+                        project_errors: p.errors.clone(),
+                        kind: CompilationKind::Notebook { project },
+                    };
+                }
+
+                (
+                    SourceMap::new(buildable_program.user_code.sources, None),
+                    buildable_program.user_code_dependencies,
+                    buildable_program.store,
+                )
+            }
+            Some(p) => {
+                trace!("compilation failure for project: {:?}", p.errors);
+                let (_, store) = qsc::compile::package_store_with_stdlib(target_profile.into());
+                return Self {
+                    package_store: store,
+                    user_package_id: 2.into(), // We don't have a legitimate user package id here, so fall back to a guess.
+                    compile_errors: Vec::new(),
+                    project_errors: p.errors.clone(),
+                    kind: CompilationKind::Notebook { project },
+                };
+            }
+            None => {
+                trace!("compiling stdlib only");
+                let (std_id, store) =
+                    qsc::compile::package_store_with_stdlib(target_profile.into());
+                (SourceMap::default(), vec![(std_id, None)], store)
+            }
+        };
+
         trace!("compiling notebook");
-        let mut compiler = Compiler::new(
-            SourceMap::default(),
+        let mut compiler = match Compiler::new(
+            sources,
             PackageType::Lib,
             target_profile.into(),
             language_features,
             store,
-            &[(std_id, None)],
-        )
-        .expect("expected incremental compiler creation to succeed");
+            &dependencies,
+        ) {
+            Ok(compiler) => compiler,
+            Err(errors) => match &project {
+                Some(p) => {
+                    trace!("compilation failure for sources: {:?}", errors);
+                    let (_, store) = qsc::compile::package_store_with_stdlib(target_profile.into());
+                    return Self {
+                        package_store: store,
+                        user_package_id: 2.into(), // We don't have a legitimate user package id here, so fall back to a guess.
```

Comment:
I thought the user's package id was standardized to a particular value, right?

Reply:
No, it's not, unfortunately. Even without external depednencies, there was the option for the user to skip the stdlib (though technically we haven't exposed it out of the compiler yet). So core is always 0, if stdlib is included it's 1, and the user codee is then either 1 or 2. But with external dependencies, those must get compiled and added to the store before the user code, and there can be any number of those (since it also includes dependencies of dependencies, and so on). So the user code could have any id greater than 0 now. To make things more confusing, when the user code fails to compile, `Compiler::new` does not add it to the store AND consumes the store, which is why I have to generate a new one here. Since it was never added to the store, it doesn't have any id at this point in the code.

Reply:
Maybe we should change the logic to always include a package 1, and in normal compilation that will be the stdlib, but if that is skipped in a compilation, then we put an empty package there. That way the user code is _always_ 2. Does this seem like a good idea? If so, I could make a git issue for this.

Reply:
No, that wouldn't work either because of dependencies. If the user code depends on Package A, which in turn depends on Package B, B must be compiled first, then A, then the user code. So the user code will be 4 by necessity. Have more dependencies, and the numbers just get higher.

Reply:
I see, thanks for the explanation.